### PR TITLE
EdgeAgent: Add handler to restart module on demand

### PR DESCRIPTION
**Pull request by varunpuranik**
Add support to allow restarting a module by invoking a direct method on the EdgeAgent.

### CODE REVIEW

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/commands/NullCommandFactory.cs [hunk 1]
```diff
@@ -23,7 +23,7 @@ public class NullCommandFactory : ICommandFactory
 
         public Task<ICommand> StopAsync(IModule module) => Task.FromResult<ICommand>(NullCommand.Instance);
 
-        public Task<ICommand> RestartAsync(IModule module) => Task.FromResult<ICommand>(NullCommand.Instance);
```
Comment (myagley) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/commands/NullCommandFactory.cs]: Instead of changing the interface, is it possible to create an `IModule` from just it's `id`?

Comment (varunpuranik) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/commands/NullCommandFactory.cs]: Emm, and what would the remaining values be? Null? 
I actually felt requiring an IModule for Restart was not necessary. Same probably goes for Stop as well.

Comment (varunpuranik) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/commands/NullCommandFactory.cs]: I can keep the interface the same and solve it in a diff way...

Comment (varunpuranik) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/commands/NullCommandFactory.cs]: Done

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/requests/RestartRequestHandler.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            RestartRequest payload = payloadOption.Expect(() => new ArgumentException("Request payload not found"));
+            if (ExpectedSchemaVersion.CompareMajorVersion(payload.SchemaVersion, "logs upload request schema") != 0)
+            {
+                Events.MismatchedMinorVersions(payload.SchemaVersion, ExpectedSchemaVersion);
+            }
+
+            Events.ProcessingRequest(payload);
+
+            DeploymentConfigInfo deploymentConfigInfo = await this.configSource.GetDeploymentConfigInfoAsync();
+            IEnvironment environment = this.environmentProvider.Create(deploymentConfigInfo.DeploymentConfig);
```
Comment (myagley) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/requests/RestartRequestHandler.cs]: Do we need a provider for this or can we get a reference to the `IEnvironment` directly through the constructor?

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/requests/RestartRequestHandler.cs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            ModuleSet modules = await environment.GetModulesAsync(cancellationToken);
+            if (!modules.TryGetModule(payload.Id, out IModule module))
+            {
+                throw new InvalidOperationException($"Module {payload.Id} not found in the current environment");
```
Comment (myagley) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/requests/RestartRequestHandler.cs]: What does an invalid operation exception return to the caller (http status code)

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/requests/RestartRequestHandler.cs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+            }
+
+            if (!(module is IRuntimeModule runtimeModule) || runtimeModule.RuntimeStatus != ModuleStatus.Running)
+            {
+                throw new InvalidOperationException($"Module {payload.Id} cannot be restarted since it is not running");
```
Comment (myagley) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/requests/RestartRequestHandler.cs]: I think if the module was stopped, I would expect this call to start it not error?

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/requests/RestartRequestHandler.cs [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+            this.configSource = Preconditions.CheckNotNull(configSource, nameof(configSource));
+            this.commandFactory = Preconditions.CheckNotNull(commandFactory, nameof(commandFactory));
+        }
+
+        public override string RequestName => "RestartModule";
+
+        protected override async Task<Option<object>> HandleRequestInternal(Option<RestartRequest> payloadOption, CancellationToken cancellationToken)
+        {
+            RestartRequest payload = payloadOption.Expect(() => new ArgumentException("Request payload not found"));
+            if (ExpectedSchemaVersion.CompareMajorVersion(payload.SchemaVersion, "logs upload request schema") != 0)
```
Comment (myagley) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/requests/RestartRequestHandler.cs]:
```suggestion
            if (ExpectedSchemaVersion.CompareMajorVersion(payload.SchemaVersion, "restart module request schema") != 0)
```

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/modules/TwinConfigSourceModule.cs [hunk 6]
```diff
@@ -77,22 +78,58 @@ protected override void Load(ContainerBuilder builder)
             builder.Register(
                 async c =>
                 {
-                    var logsUploader = c.Resolve<ILogsUploader>();
-                    var runtimeInfoProviderTask = c.Resolve<Task<IRuntimeInfoProvider>>();
-                    var logsProviderTask = c.Resolve<Task<ILogsProvider>>();
-                    IRuntimeInfoProvider runtimeInfoProvider = await runtimeInfoProviderTask;
-                    ILogsProvider logsProvider = await logsProviderTask;
+                    await Task.CompletedTask;
```
Comment (myagley) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/modules/TwinConfigSourceModule.cs]: Why is this needed?

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/requests/RestartRequestHandler.cs (hunk 5 above)
Comment (varunpuranik) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/requests/RestartRequestHandler.cs]: Oops, thanks

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/modules/TwinConfigSourceModule.cs (hunk 6 above)
Comment (varunpuranik) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/modules/TwinConfigSourceModule.cs]: It is not.. forgot to cleanup

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/requests/RestartRequestHandler.cs (hunk 3 above)
Comment (varunpuranik) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/requests/RestartRequestHandler.cs]: 500.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/requests/RestartRequestHandler.cs (hunk 4 above)
Comment (varunpuranik) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/requests/RestartRequestHandler.cs]: Emm, I thought in that case we should let EdgeAgent start it (as per backoff policy, etc)?

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/requests/RestartRequestHandler.cs (hunk 2 above)
Comment (varunpuranik) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/requests/RestartRequestHandler.cs]: The thing is IEnvironment needs DeploymentConfig, which can get updated, so we cannot use one created initially.

Comment (varunpuranik) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/requests/RestartRequestHandler.cs]: I think this whole thing needs to be refactored/restructured. But in another PR...

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/requests/RestartRequestHandler.cs (hunk 4 above)
Comment (arsing) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/requests/RestartRequestHandler.cs]: It does make sense to think of restarting as "Ensure it's stopped, then ensure it's started". So for an already stopped module it would just start it.

Comment (varunpuranik) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/requests/RestartRequestHandler.cs]: Actually Mike and I agreed to not check the status at all.. simply invoke Restart on the module. That would be similar to what iotedge does.

Comment (arsing) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/requests/RestartRequestHandler.cs]: Cool. That should have the same behavior then (via `docker restart`).

Comment (myagley) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/requests/RestartRequestHandler.cs]: Should this if check be removed then?
